### PR TITLE
Replace mod_perl2 with Mojolicious

### PR DESCRIPTION
**Pull request by drgrice1**
It is time to say goodbye to mod_perl2, and unbind webwork2 from apache2.  With Mojolicious there are many more deployment options.

The WeBWorK::Cookie module which was really just an alias for the CGI::Cookie module has been removed.  Instead this just uses the Mojolicious cookie methods.  Note that this does change the way the $CookieLifeTime setting works.  Previously it was a string representation of the life time like "+7d", and now it must be numeric and is always in seconds or the string "session" (in which case the expires and max-age values are not set and the cookie expires with the session).

Note that configuration that was previously in webwork2.apache2.4.dist.conf is now in webwork2.mojolicious.dist.yml. Unless you want to change the settings in that file, you don't need to copy that file to webwork2.mojolicious.yml.  The `dist` file will be used directly if the webwork2.mojolicious.yml file does not exist.

The XMLRPC::Lite and Apache::XMLRPC::Lite modules are not longer used.  The xml endpoints have been replaced with a direct remote procedure call endpoint.  This removes the XMLRPC middle man.

Note that the paths for this have changed.

The html2xml path still exists, but is an alias for the new render_rpc path.  The html2xml name doesn't make sense anymore since it does not use xml at all.  It should be considered deprecated and not used in new code.  It is provided to give those that use it time to make the change.   It still works the same at it did before.

The instructorXMLHandler path is now instructor_rpc.  Note that the response that is returned is now JSON, instead of a string that is parsed into JSON. If there is an error then the return object will contain an "error" key explaining the error.  There is no backwards compatibility path for this.  I don't think that this path was really used by anyone other than webwork2 itself.  If I am mistaken, an alias can be made.

These remote procedure call endpoints now use the usual authentication paths in the WeBWorK.pm dispatch method instead of waiting until the WebworkWebservice.pm module.  For this to work it is neccessary to change to using the usual authentication parameters.  Now use `user` instead of `userID`, `passwd` instead of `course_password`, and `key` instead of `session_key`.  Also, `courseName` is obsolete.  Previously both `courseID` and `courseName` were allowed.  However, if you use the now deprecated /webwork2/html2xml endpoint, you can still use the old parameters.  Those will not work with /webwork2/render_rpc.  You should work to switch to the new /webwork2/render_rpc endpoint, as the /webwork2/html2xml endpoint will be deleted eventually.

Note that the forcePortNumber parameter is removed.  Instead the port should be included in the $server_root_url in site.conf.

Note that the badly broken, previously disabled, and probably never completely written WebworkWebservice::MathTranslators package has been deleted.  All of the methods in the other WebworkWebservice packages now actually work.

Also a global javascript `webworkConfig` variable is now injected into the templates (in system.template and gateway.template) that provides system information for javascript files.  At this point it contains the single key `webwork_url`.  This is the value of $webwork_url in site.conf (typically /webwork2).  Although it is actually the empty  string if $webwork_url is set to '/'.  This makes that setting actually work again.

Note that the SOAP endpoints still work.  They are disabled by default as they were before.  To enable them, uncomment the `soap_authen_key` parameter in `conf/webwork2.mojolicious.yml` and change it to a secure value.

The docker build has been updated for this.  The build has also been updated to use Ubuntu 22.04.  Note that apache2 is no longer installed in the container.  Instead webwork2 is served via hypnotoad directly run as the www-data user.

For developers a script has been added to run the morbo dev server.  It is `./bin/dev_scripts/webwork2-morbo`.  Note that the script is actually the Mojolicious `morbo` script, but it locates the `bin/webwork2` script and runs it as well as adds some convenient "watch" directories (webwork2/htdocs/js, webwork2/htdocs/themes, webwork2/conf, pg/lib, and pg/htdocs).  Note you don't want to watch all of the webwork/htdocs directory or a file written in the htdocs/tmp directory will cause morbo to reload and end the request writing the file.  Other than not needing to pass the bin/webwork2 script argument, the script accepts all other arguments that morbo does and works the same.  You can add more watch directories or files using the `--watch` parameter.  You can change the port with `--listen=http://*:4000`.  Note that you must include the port with the $server_root_url in site.conf when running with morbo.  I use
```perl
$server_root_url =
	(exists $ENV{MOJO_MODE} && (!$ENV{MOJO_MODE} || $ENV{MOJO_MODE} ne 'production'))
	? 'http://localhost:3000'
	: 'http://localhost';
```
to make site.conf work for both morbo and with hypnotoad with the production deployment.

For deployment by proxy via apache2 there are two files in webwork/conf that can be used.  Those are conf/webwork2.apache2.4.dist.conf and conf/webwork2.dist.service.  The files may need modification in some cases.  Copy conf/webwork2.apache2.4.dist.conf to conf/webwork2.apache2.conf and link that to /etc/apache2/conf-enabled/webwork2.conf just as was done with conf/webwork.apache2.4-config before.  Note that you will need the apache2 "proxy" and "proxy_http" modules enabled (and maybe "headers" and "rewrite"?).  Of course restart apache2 after that.  Copy conf/webwork2.dist.service to conf/webwork2.service and run `sudo systemctl enable $PWD/conf/webwork2.service` and then `sudo systemctl start webwork2` to start the hypnotoad server in production mode.

I am sure there are more things that will need to be done, but this takes care of most of the transition from mod_perl2 to Mojolicious.

I forgot to mention the perl module changes.  You no longer need Apache2::Request, Apache2::ServerRec, and Apache2::ServerUtil (or any of mod_perl2), and you also don't need CGI::Cookie or XMLRPC::Lite.  You need to add Mojolicious, Mojolicious::Plugin::NotYAMLConfig, Future::AsyncAwait, and Data::Structure::Util.  On Ubuntu you can install `libmojolicious-perl`, `libfuture-asyncawait-perl`, and `libdata-structure-util-perl`.  You need to install Mojolicious::Plugin::NotYAMLConfig from cpan though as that is not packaged for Ubuntu.

Also remember to run `./htdocs/generate-assets` (or just `npm install`) in the webwork2 directory as there have been javascript changes.

### CODE REVIEW

**Comment by drgrice1** [comment]
@drdrew42 and @pstaabp:  You should be able to get this working quite easily with morbo.  You should have all the necessary modules already installed.  Remember to add the port in site.conf.

**Comment by somiaj** [comment]
Running some initial tests and things seem to work. I am able to get my development server up and running, and webwork2 seems to work as expected (checked a few pages, tested the library browser,  manual grader, and problem editor). Though I'm a little confused on how to configure the port correctly. I have my `$server_root_url  = 'http://192.168.122.154:80';` and when I run `bin/webwork2 daemon` or `bin/dev_scripts/webwork2-morbo` the server runs on `http://localhost:3000`. I can add the `--listen=http://*:80` option to ` bin/dev_scripts/webwork2-morbo` to get the server to run on port 80, but if I add that option to `bin/webwork2` I get invalid option. When running `bin/webwork2 daemon` I see the following:

```
[2022-10-20 22:03:31.34189] [5028] [info] The webwork url on this site is http://192.168.122.154:80/webwork2
[2022-10-20 22:03:31.36693] [5028] [info] Listening at "http://*:3000"
Web application available at http://127.0.0.1:3000
```

File: conf/webwork2.mojolicious.dist.yml [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+# Uncomment the line below to use the WebworkSOAP handlers.  Make sure that you
+# change the value to something more secure.  Note that it must consist entirely
+# of digits.
+#soap_authen_key: 123456789123456789
+
+# If a remote site is using this WW server to process WW problems and embed them
+# into web pages, the web browser might not permit the WW problems to be
+# displayed, as it violates security policy to mix content from different
+# originis without an explicit directive from the server that it is safe.
```
Comment (taniwallach) [inline conf/webwork2.mojolicious.dist.yml:41]: Typo: originis

Comment (drgrice1) [reply conf/webwork2.mojolicious.dist.yml]: Thanks @taniwallach.  I cut and pasted that the now deleted conf/webwork.apache2.4-config.dist file.  The typo was there already.  I fixed it.

**Comment by drgrice1** [comment]
@somiaj:  Generally you don't directly run the `bin/webwork2` script.  Usually you run the app via `morbo` or `hypnotoad`.  `morbo` is intended for development use, and usually you don't use port 80 for that although you can.  `hypnotoad` is what you should use for production use, and for that you set the port in `conf/webwork2.mojolicious.yml`.

**Comment by drgrice1** [comment]
@somiaj:  You can set the port when running `./bin/webwork2` directly with `./bin/webwork2 daemon -l http://*:5000` or `./bin/webwork2 daemon --listen=http://*:5000`.  You probably won't be able to use port 80 for this unless you run the script as the root user due to the usual linux restrictions.

File: lib/Mojolicious/WeBWorK.pm [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+			$r->any('/webwork2_wsdl')->to('SOAP#wsdl');
+			$r->post('/webwork2_rpc')->to('SOAP#dispatch');
+		} else {
+			$app->log->info(qq{Invalid soap_authen_key "$config->{soap_authen_key}". }
+					. 'It must consist entirely of digits.  SOAP endpoints NOT enabled.');
+		}
+	}
+
+	# Send all routes under $webwork_url to the handler.
+	# Note that these routes must come last to support the cast that $webwork_url is '/'.
```
Comment (taniwallach) [inline lib/Mojolicious/WeBWorK.pm:147]: Typo: "cast"

Comment (drgrice1) [reply lib/Mojolicious/WeBWorK.pm]: Fixed.

**Comment by taniwallach** [comment]
@drgrice1 Thanks for working on this. Getting rid of the dependency on mod_perl2 will certainly already make it far simpler to keep webwork2 viable on Red Hat derived distributions, and dropping the dependency on this out-dated platform should help with other things over time. I tried to get a bit of an idea of how things are being changed by scanning the changes, but it will take some effort to get a real handle on the changes.

---

This week I was asked to look into whether the multi-institution server I help operate could possible support SAML based authentication (for use by an institution which cannot use LTI). As a result, I am a bit more aware of the additional authentication options currently supported by webwork2, and in particular on the option of using Shibboleth.

It seems that one side effect of the planned changes will be that the Shibboleth and Cosign authentication may no longer work, unless they are handled at the level of an Apache proxy server, as from what I understand from the code and the Wiki documentation, these authentication approached depend on Apache doing the actual authentication process, and then providing relevant data to WeBWorK.

I would not let that issue hold back these changes, but we may want to reach out to people who depend on Shibboleth and Cosign to help find suitable solutions for using those approaches with a Mojolicious based webwork2.

**Comment by drgrice1** [comment]
Yes, some of those authentication approaches will need to be figured out.  Most likely those will need to set things up via the apache2 proxy deployment method and not the direct hypnotoad deployment method.

**Comment by pstaabp** [comment]
Overall, looks great and runs much speedier on my develop machine. 

In the Library Browser, the problems aren't rendering correctly.  In the morbo output, there are 304 errors on css/js files. 

```
[2022-10-25 07:07:49.36659] [65357] [trace] [_9rQdtI8ISu_] GET "/webwork2_files/themes/math4/math4.fc51f57d.min.css"
[2022-10-25 07:07:49.36798] [65357] [trace] [_9rQdtI8ISu_] Routing to a callback
[2022-10-25 07:07:49.36994] [65357] [trace] [_9rQdtI8ISu_] 304 Not Modified (0.003331s, 300.210/s)
[2022-10-25 07:07:49.37643] [65357] [trace] [CaBzmlDHUiaC] GET "/webwork2_files/themes/math4/bootstrap.f038b6ee.min.css"
[2022-10-25 07:07:49.37745] [65357] [trace] [CaBzmlDHUiaC] Routing to a callback
[2022-10-25 07:07:49.37850] [65357] [trace] [CaBzmlDHUiaC] 304 Not Modified (0.002062s, 484.966/s)
[2022-10-25 07:07:49.38188] [65357] [trace] [K2ByKrjI_HW7] GET "/webwork2_files/themes/math4/math4-overrides.f4c96707.min.css"
[2022-10-25 07:07:49.38303] [65357] [trace] [K2ByKrjI_HW7] Routing to a callback
[2022-10-25 07:07:49.38401] [65357] [trace] [K2ByKrjI_HW7] 304 Not Modified (0.002127s, 470.146/s)
[2022-10-25 07:07:49.38697] [65357] [trace] [wtHjuTd1mYpN] GET "/webwork2_files/js/apps/MathJaxConfig/mathjax-config.d097a5f4.min.js"
[2022-10-25 07:07:49.38823] [65357] [trace] [wtHjuTd1mYpN] Routing to a callback
[2022-10-25 07:07:49.38982] [65357] [trace] [wtHjuTd1mYpN] 304 Not Modified (0.002815s, 355.240/s)
[2022-10-25 07:07:49.39167] [65357] [trace] [xhh8U6pRsQf2] GET "/webwork2_files/js/apps/SetMaker/setmaker.0348ac68.min.js"
[2022-10-25 07:07:49.39248] [65357] [trace] [xhh8U6pRsQf2] Routing to a callback
[2022-10-25 07:07:49.39422] [65357] [trace] [xhh8U6pRsQf2] 304 Not Modified (0.002523s, 396.354/s)
[2022-10-25 07:07:49.39645] [65357] [trace] [4DlBGKjFkHi0] GET "/webwork2_files/themes/math4/math4-overrides.90695a18.min.js"
[2022-10-25 07:07:49.39837] [65357] [trace] [4DlBGKjFkHi0] Routing to a callback
[2022-10-25 07:07:49.40470] [65357] [trace] [4DlBGKjFkHi0] 304 Not Modified (0.00823s, 121.507/s)
[2022-10-25 07:07:49.40731] [65357] [trace] [1oO6D0LdJVGE] GET "/webwork2_files/themes/math4/math4.04d32dc9.min.js"
[2022-10-25 07:07:49.40884] [65357] [trace] [1oO6D0LdJVGE] Routing to a callback
[2022-10-25 07:07:49.41053] [65357] [trace] [1oO6D0LdJVGE] 304 Not Modified (0.00302s, 331.126/s)
```

Both math and images are not showing and the font is not the same. 

I did have to change some permission around since morbo is now running as a regular user, but don't see any permissions problems.

File: bin/dev_scripts/webwork2-morbo [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+die extract_usage if $help;
+
+my $webwork_root = curfile->dirname->dirname->dirname->to_string;
+
+# Load the configuration file to obtain the PG root directory.
+my $config_file = "$webwork_root/conf/webwork2.mojolicious.yml";
+$config_file = "$webwork_root/conf/webwork2.mojolicious.dist.yml" unless -e $config_file;
+my $config = LoadFile($config_file);
+
+# Add directoriess to the watch list that might change frequently while developing for webwork2 and that need the server
```
Comment (pstaabp) [inline bin/dev_scripts/webwork2-morbo:79]: typo: directoriess to directories

File: conf/README.md [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+* To run the server without ssl certificates change `listen` in the `hypnotoad` section at the end of the file to
+  `- http://*:80`.  This is not recommended for production use.
+* To use ssl certificates change `listen` in the `hypnotoad` section to
+  `- https://*:443?cert=/path/to/fullchain.pem&key=/path/to/privkey.pem`.
+* Change `proxy: 1` to `proxy: 0` or comment out that line.
+* You may also want to adjust the other settings in the file.  The `server_root_url_redirect` setting may be useful.
+* Instead of using that setting, you can also copy `htdocs/index.dist.html` to `htdocs/index.html` and that will be the
+  server front page.
+
+Then set up the systemd service:
```
Comment (pstaabp) [inline conf/README.md:93]: Are these instructions for ubuntu only?  That is, does RedHat or other dists use systemd and systemctl?

**Comment by drgrice1** [comment]
@pstaabp:  An http response code of 304 is not an error.  That just means your browser has cached the file and is asking if it needs to retrieve it again.  The server is responding with a 304, telling your browser, "No, you don't need a new version.  What you have is good!"

File: bin/dev_scripts/webwork2-morbo (hunk 3 above)
Comment (drgrice1) [reply bin/dev_scripts/webwork2-morbo]: Fixed.

File: conf/README.md (hunk 4 above)
Comment (drgrice1) [reply conf/README.md:93]: The instructions are generally specific to Ubuntu, but should be adaptable to other distributions.  Other distributions do use systemd.

**Comment by drgrice1** [comment]
@pstaabp:  You say that problems are not rendering correctly in the library browser.  Did you update your javascript files in htdocs for webwork2 with either `./htdocs/generate-asses` or `npm install` in the htdocs directory?

**Comment by pstaabp** [comment]
> Did you update your javascript files in htdocs for webwork2 with either ./htdocs/generate-asses or npm install in the htdocs directory?

I did.  I'm getting errors of the form: `Failed to load resource: net::ERR_CONNECTION_REFUSED.` for items like: `http://localhost/pg_files/js/apps/Problem/problem.eadaf2d8.min.css`

I'm running on port 3000 (didn't change any of the standard options when running `webwork2-morbo`.  It seems like calls to  `render_rpc` don't have a port set, so I assume is trying port 80.  This may be what the issue is.

**Comment by pstaabp** [comment]
Just reread your instructions.  It was key to add
```
$server_root_url =
	(exists $ENV{MOJO_MODE} && (!$ENV{MOJO_MODE} || $ENV{MOJO_MODE} ne 'production'))
	? 'http://localhost:3000'
	: 'http://localhost';
```
to `site.conf`.

**Comment by drgrice1** [comment]
Yeah, that was going to be my next question.  The port must be included in the `$server_root_url`.

**Comment by pstaabp** [comment]
Just spent a bit more time going through everything.  I'm logging in with the basic authentication and running (via morbo) through all of the pages seems to work well now. 

Seems like the big thing will be testing in production mode with all of the nuances (LTI, authentication methods, ...) will be the tricky thing with this.

**Comment by drgrice1** [comment]
You can remove LTI authentication from that list.  That works fine with this.

**Comment by pstaabp** [comment]
I'm trying to get this working on a ubuntu server.  Just noticed that installing `mojolicious-perl` has `libfuture-asyncawait-perl` as a dependency, so it will be installed automatically.

**Comment by drgrice1** [comment]
The package on Ubuntu for Mojolicious is `libmojolicious-perl`.  `libfuture-asyncawait-perl` is not a dependency of the `libmojolicious-perl` package, but it is recommended by the package.  So it is installed if you ask for recommended packages.  Recommended packages may be installed by default, depending on your system configuration.

Note also that both of those packages for Ubuntu 20.04 or before will not work because the versions of Mojolicious and Future::AsyncAwait are not new enough.  So for those versions of Ubuntu, you will need to install from cpan.  Ubuntu 22.04 has new enough versions, and the packages work fine.

**Comment by somiaj** [comment]
Reporting another successful test on a virtual machine server behind an apache proxy, and so far no issues. My setup is to connect to the proxy via https, then the proxy connects to mojolicious on port 3000 (just used the default). Basic authentication, email, homework problems, manual grader, library browser, are all running as expected. For what it is worth, I'm running this on Debian bullseye (current stable) and had to backport 6 packages (I just prefer this over CPAN if I can) to get the versions needed for libmojolicious-perl and libfuture-asyncawait-perl, but outside of that no issues.

**Comment by pstaabp** [comment]
Trying some other day-to-day stuff with this.  In the library browser (SetMaker) if you "Add All" problems to a set, I'm getting the following: `Course 'instructorXMLHandler' not found: No such file or directory at /opt/webwork/webwork2/lib/WeBWorK.pm line 276.`

That line doesn't make any sense, but the following is in the morbo output: 
```
[2022-10-29 06:54:09.94711] [78542] [trace] [YZKnMuGytdP9] POST "/webwork2/instructorXMLHandler"
[2022-10-29 06:54:09.94829] [78542] [trace] [YZKnMuGytdP9] Routing to controller "Mojolicious::WeBWorK::Controller::Handler" and action "handler"
[2022-10-29 06:54:09.96947] [78542] [error] [YZKnMuGytdP9] [5cb15bc7-d42b-5be7-bb5c-ee249e546918::0463276c-5778-11ed-bd70-9911f22587e5] [http://localhost:3000/webwork2/instructorXMLHandler] {"HTTP Headers":{"Connection":"keep-alive","Referer":"http://localhost:3000/webwork2/fitchburg_problems/instructor/setmaker/","Accept-Language":"en-US,en;q=0.9","Content-Length":"233","User-Agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/16.1 Safari/605.1.15","Accept-Encoding":"gzip, deflate","Origin":"http://localhost:3000","Content-Type":"application/x-www-form-urlencoded;charset=UTF-8","Accept":"*/*","Cookie":"WeBWorKCourseAuthen.fitchburg_problems=peter%09qTsBJTcPd19b8glyGc94LHkrpRqeZXDL%091667040836; WeBWorKCourseAuthen.admin=admin%09S4pCOEFbnPCSt8k93z1Q5q68dOmHh7Of%091667040761; WeBWorKCourseAuthen.staab_course=peter%09WcYyirA46e415qnY9g5kofOUcj9JPfwC%091667040601; username-localhost-8888=\"2|1:0|10:1666893429|23:username-localhost-8888|44:OTcwOGI3YTIzYTI2NDE3MTlhMDc5ZDQyYWU2MjdkODU=|17c8eb25a2a3d5470e3c36b8f1045a1924b925421b62183e65e266b192c982e2\"; username-localhost-8889=\"2|1:0|10:1666893261|23:username-localhost-8889|44:NWZkMTUxYTlmOTVmNDQzMWE0OGFmMThlNjhiMWZmOTQ=|24569de6a91fd57352371982ca082c41f24542d303e989c3ebfdb08c5cc7ba46\"; username-localhost-8890=\"2|1:0|10:1666779627|23:username-localhost-8890|44:OWYyZGZjY2I3ODBiNDk5ZGFhZDc5YjNkMTExYTIyNGM=|a2504f0c6afb01e08d8f2a3aa9c0b10320b0f3ba5bfb4e581bfe4f9c1c4e161e\"; username-localhost-8891=\"2|1:0|10:1666013632|23:username-localhost-8891|44:NDZjZjdmZDg5ZmE1NDg5NzkyNjhmZWVhYTQ5YWMxMjk=|27d6999229fbc264ba3480f9b9d033859defd14c367221e86e0082c23a82b5b6\"; _xsrf=2|8c08a280|ad204e3d5feeb69d17cc86ea1acfe76e|1665428281","Host":"localhost:3000"},"Error record identifier":"5cb15bc7-d42b-5be7-bb5c-ee249e546918::0463276c-5778-11ed-bd70-9911f22587e5","Method":"POST","Warnings":[],"Time":"Sat Oct 29 06:54:09 2022","URI":"http://localhost:3000/webwork2/instructorXMLHandler"} Course 'instructorXMLHandler' not found: No such file or directory at /opt/webwork/webwork2/lib/WeBWorK.pm line 276.
 * in Future::result called at line 921 of /Users/pstaab/perl5/perlbrew/perls/perl-5.30.3/lib/site_perl/5.30.3/Future.pm
[2022-10-29 06:54:09.97068] [78542] [trace] [YZKnMuGytdP9] 403 Forbidden (0.023594s, 42.384/s)
```

**Comment by pstaabp** [comment]
Never mind, I didn't `npm install` to update the js files.  It was using `setmaker.js` from develop.

**Comment by pstaabp** [comment]
However, I am getting the error in `SetMaker` switching target problem sets, a bootstrap dialog pops up with the error: `WebworkWebservice: Invalid command listSetProblems`

**Comment by drgrice1** [comment]
@pstaabp:  I suspected the `npm install` issue again on the instructorXMLhandler.  I went to check on that though, and saw the invalid command `listSetProblems`.  That command is now `listGlobalSetProblems`.  I changed it because that is what the actual method called is.  There is now a one-to-one correspondence to commands and WebworkWebservice methods.  However, I forgot to update setmaker.js.  That is fixed now, so update your local clone.

**Comment by drgrice1** [comment]
Also run `npm install` (or just `./htdocs/generate-assets.js`) again.  I recommend using `./htdocs/generate-assets.js` instead of `npm install`.  Yes, `npm install` is more thorough, but unless the `package.json` file changes, there is no need.

**Comment by pstaabp** [comment]
Another run of `generate-assets.js` has done the trick.

**Comment by drgrice1** [comment]
I updated the instructions in `conf/README.md` for direct deployment to deploy hypnotoad with a user that is not the root user or any user that has a login account.  The method uses `authbind` which should be available on most linux distributions.  This seems to be the most secure way of doing this that I have found.  It allows the chosen user (you can use `www-data` on Ubuntu for this) to bind to a low numbered port (like 80 or 443), and is quite simple to set up.

**Review by drdrew42 — APPROVED** [review]
This is HYUUUUGE 🥳

**Review by pstaabp — APPROVED** [review]

**Comment by drgrice1** [comment]
I have now completely removed the WeBWorK::Authen::RPC module.  Remote procedure calls really do just use the usual authentication.  There is no need to disable cookies.

@pstaabp:  The xmlrpc instances that you noticed in the authen modules have been fixed.

**Comment by drgrice1** [comment]
Actually, cookies do need to be disabled for RPC calls.  However, that is now done in the usual WeBWorK::Authen module for RPC calls instead of in a separate module.  This makes it possible to make RPC calls using the other authentication modules as well.

**Comment by Alex-Jordan** [comment]
When serving without apache, is there a need to worry about memory leak issues that are currently addressed with apache config settings? The two things I know of are the `MaxRequestWorkers` setting for the prefork module and use of `Apache2::SizeLimit`. With hypnotoad, are there counterparts to these safeguards?

**Comment by somiaj** [comment]
@Alex-Jordan check out the `webwork2.mojolicious.yml` configuration file. You can configure workers, and other such things there. Though I'm not sure exactly how the settings relate, I don't see a size limit listed for instance.

**Pull request merged**